src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -35,6 +35,215 @@ public static partial class Base64
         ///   or if the input is incomplete (i.e. not a multiple of 4) and <paramref name="isFinalBlock"/> is <see langword="true"/>.
         /// </returns>
         public static unsafe OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> bytes, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true)
+        {
+            // Validation must occur prior to decoding as the actual length will impact future calculations
```

Comment:
This will most likely cause a perf-regression, as decoding changed from $O(n)$ to be $O(2n)$ with this upfront check. If there's space encountered it's even $O(3n)$ due the "copy and remove space" loop.

Maybe it's better to just start decoding w/o any upfront check or copy, and if an invalid input is encountered fall back and decide what to do. If it's truly invalid then exit. If it's space, then ignore that position and continue. Thus the $O(n)$ remains.

To implement this you can split the decoding into two methods. One "driver" method, and one "worker" method. This is only relevant for the invalid-case, as then the driver can evaluate the state and call the worker again. Look at the buffer-chain examples / tests to see how someting like this can be done.